### ISSUE TITLE
Franzininho WiFi Board (ESP32-S2)

### ISSUE THREAD

**Issue by arcostasi** [wokwi/wokwi-features#196]
Franzininho WiFi Board is a development board to evaluate ESP32-S2 Modules (Wroom and Wrover) and develop the new generation of Franzininho Boards. The Franzininho project was created to develop skills in people in the areas of electronics and programming, through activities in the DIY format and in conjunction with maker culture in Brazil.

Project: https://github.com/Franzininho/Franzininho-WIFI

The project is an open-source hardware and is available on CERN open hardware license.

The Franzininho WiFi board is certified in OSHWA: [UID BR000006](https://certification.oshwa.org/br000006.html)

**Comment by urish** [comment]
Great to see this racing to the top of the [feature list](https://wokwi.com/features)!

Preparing the data for the board design:
- [Pinout](https://www.embarcados.com.br/wp-content/uploads/2021/09/pinagem-franzininho-wifi.png)
- [Board dimensions](https://github.com/Franzininho/Franzininho-WIFI/blob/main/Franzininho%20WIFI/pdf/franzininhoWIFI-pcb.pdf)

Is there any publicly available SVG drawing of the board?

**Comment by urish** [comment]
Initial version of the board is now available in [wokwi/wokwi-boards](https://github.com/wokwi/wokwi-boards/tree/master/boards/franzininho-wifi). 

You can go ahead and load it in the simulator by following the [instructions for loading a custom board](https://github.com/wokwi/wokwi-boards#testing-a-custom-board). 

You'll be able to preview the board and draw connections, but the simulation does not work yet:
1. The build will fail due to espressif/arduino-esp32#5735
2. The ESP32-S2 chip simulation is still WiP

**Comment by arcostasi** [comment]
Thanks Uri, any results with the armel toolchain mentioned in https://github.com/espressif/arduino-esp32/issues/5735#issuecomment-937519380

**Comment by arcostasi** [comment]
> Is there any publicly available SVG drawing of the board?

Sorry I didn't see the question, but we love the svg created by Ariella.

**Comment by urish** [comment]
> Thanks Uri, any results with the armel toolchain mentioned in espressif/arduino-esp32#5735 (comment)

Yes, positive results! I just added a comment there

**Comment by urish** [comment]
Some more updates: working to get the ESP32-S2 simulator to a usable state, then integrate it with Wokwi.

So far, was able to get MicroPython 1.17 prompt:

```
$ npx ts-node demo/micropython

MicroPython v1.17-74-gd42cba0d2-dirty on 2021-10-09; ESP32S2 module with ESP32S2
Type "help()" for more information.
>>> print("Hello, world!", 3**6)
Hello, world! 729
>>> import time
>>> time.ticks_ms()
945648215
>>> 
```

So this is a good sign. It means that the following peripherals are working correctly:
- CPU core
- UART
- Flash access (otherwise MicroPython fails to start with FS initialization errors)
- System timer (partially - we shouldn't get a ticks_ms() value this large when we just started)

There's still much work to do before it'll be usable:
- USB (as Franzininho WiFi uses USB over UART)
- GPIO (so we can eventually see the LED blinking!)
- Fix the timers (so delay() works reliably)

And probably a bunch of other things we'll discover along the way :-)

**Comment by arcostasi** [comment]
What great news Uri, there really is a lot of work to be done! If you need, I can go testing and validating with Franzininho WiFi hardware. Very good to know about was able to get MicroPython, Franzininho WiFi now has [official support for CircuitPython](https://circuitpython.org/board/franzininho_wifi_wroom/), I believe we will explore this feature a lot. :-)

**Comment by urish** [comment]
Excellent! CircuitPython will probably be another stage, first let's get the Arduino stuff to work!

**Comment by urish** [comment]
Progress update: GPIO is functional! 

So as soon as the Arduino-ESP32 issue will be resolved, we'll be able to close the loop and have Blink working.

I'm now starting to work on USB, which will enable CircuitPython (as well as easier output from the Arduino Core).

**Comment by urish** [comment]
Video showing CircuitPython work-in-progress: https://twitter.com/WokwiMakes/status/1449804926478168068

**Comment by urish** [comment]
Now you can try it yourself!

Find the complete instructions here: https://wokwi.com/arduino/projects/313606939786347074

**Comment by urish** [comment]
Update: NeoPixel also works now!

To try, paste the following code in the CircuitPython REPL:

```python
import board
import neopixel_write
import digitalio

pin = digitalio.DigitalInOut(board.NEOPIXEL)
pin.direction = digitalio.Direction.OUTPUT
green = bytearray([255, 0, 0])
neopixel_write.neopixel_write(pin, green)
```

**Comment by urish** [comment]
Update: you can now run the [Franzininho WiFi CircuitPython Demo Project](https://wokwi.com/arduino/projects/313606939786347074) out of the box, without any special instructions.

I hope the Arduino Core issue gets fixed soon, and so we'll also have a set of demos with Arduino.

**Comment by arcostasi** [comment]
Very good, doing some tests with CircuitPython, a question, is time.sleep already implemented?

This is the code I'm testing:

```python
""" Hello, World! (Blink LED)"""
import board
import time
from digitalio import DigitalInOut, Direction

# Setting the LED pin, the LED I configured was pin 4 = IO4
led = DigitalInOut(board.IO4)
led.direction = Direction.OUTPUT

# infinite loop - always running
while True:    
    led.value = True
    time.sleep(0.5)
    led.value = False
    time.sleep(0.5)
```

Diagram link: https://wokwi.com/arduino/projects/314165931402068544

**Comment by FBSeletronica** [comment]
> Very good, doing some tests with CircuitPython, a question, is time.sleep already implemented?
> 
> This is the code I'm testing:
> 
> ```python
> """ Hello, World! (Blink LED)"""
> import board
> import time
> from digitalio import DigitalInOut, Direction
> 
> # Setting the LED pin, the LED I configured was pin 4 = IO4
> led = DigitalInOut(board.IO4)
> led.direction = Direction.OUTPUT
> 
> # infinite loop - always running
> while True:    
>     led.value = True
>     time.sleep(0.5)
>     led.value = False
>     time.sleep(0.5)
> ```
> 
> Diagram link: https://wokwi.com/arduino/projects/314165931402068544

I had the same problem here: 
![image](https://user-images.githubusercontent.com/4184762/139919791-4fdf93b4-e6a9-4808-bbf2-8480be3bcd45.png)

The led is always ON

**Comment by urish** [comment]
Yes, it seems like the internal timer is going way too fast, so the sleep finishes immediately. I'll try to figure out why this happens

**Comment by urish** [comment]
Sleep fixed!

**Comment by urish** [comment]
Progress update: Filesystem works!

You can try it here: https://wokwi.com/arduino/projects/313606939786347074

Note that the code still doesn't run automatically (CP starts in Safe mode, I still have to figure out how to avoid this), so you have to manually load it from the REPL:

```python
import code
```

**Comment by urish** [comment]
espressif/arduino-esp32#5735 fixed! Meaning you can now enjoy:

- [RGB Blink on Franziniho WiFi using the Arduino Core](https://wokwi.com/arduino/projects/314916550651937346)

There's still some issue with the USB Serial output on Arduino Core, I have to debug it and figure out why it doesn't print any output. But other than that, it's pretty usable :)

**Comment by arcostasi** [comment]
Thanks Uri, good to hear that espressif fixed this! everyone is to be congratulated 😉

**Comment by urish** [comment]
USB Serial on Arduino core works! Example:

https://wokwi.com/arduino/projects/316092312585241154

So the Arduino core is pretty much complete. Let's see what people build with it. Over the next few days, I'll add the Franzininho WiFi to the "new project" menu, documentation, etc. so it'll become generally available.

My next action item here is to make CircuitPython load the user code when the simulation starts.

**Comment by urish** [comment]
Forgot to say - also added an automated test case for Franzininiho WiFi simulation: https://github.com/wokwi/wokwi-tests/tree/master/franzininho-wifi/usb-serial

**Comment by arcostasi** [comment]
Thanks Uri, I'll share it in the community and try it out!

**Comment by urish** [comment]
Some updates: I implemented a bunch of performance speedups (similar to the ones I did for the ESP32), and now the ESP32-S2 simulation runs about 4 times faster.

Getting the WiFi to work is still a big challenge - I'll keep update here as I slowly make progress.

**Comment by arcostasi** [comment]
> Some updates: I implemented a bunch of performance speedups (similar to the ones I did for the ESP32), and now the ESP32-S2 simulation runs about 4 times faster.

That's great Uri, it's already a huge improvement! In my tests I've already noticed the improvement in speed ;)

> Getting the WiFi to work is still a big challenge - I'll keep update here as I slowly make progress.

All right, but that's already a big step forward! I'll do more tests! Thanks

**Comment by urish** [comment]
Good news: the WiFi is working!

Examples:
- [NTP Time](https://wokwi.com/arduino/projects/323043284024820308)
- [HTTP Server](https://wokwi.com/arduino/projects/323043611101889108) - Requires the [Wokwi IoT Gateway](https://docs.wokwi.com/guides/esp32-wifi#the-private-gateway)

Note: TLS support isn't there yet. We'll need to implement the crypto hardware accelerators - SHA, AES, RSA and possibly a few more before TLS/HTTPS will be functional. Similar to what we did in #267.

**Comment by arcostasi** [comment]
What good news! I did some experiments and everything worked just fine! The performance is also excellent!

Here is my test example:

- [Weather HTTP Server](https://wokwi.com/arduino/projects/323063403217158738) - Requires the [Wokwi IoT Gateway](https://docs.wokwi.com/guides/esp32-wifi#the-private-gateway)

Thanks Uri

**Comment by urish** [comment]
Very cool!

![image](https://user-images.githubusercontent.com/892318/153160393-3272cff2-a1ea-41d4-ac5a-ead080005d61.png)

Did you have any problems getting the DHT sensor to work?

**Comment by FBSeletronica** [comment]
Amazing Uri! I will test and make some videos. An article on the "Embarcados" website is also worth sharing with the Brazilian community. This is a piece of good news! Thank you.

**Comment by arcostasi** [comment]
> Did you have any problems getting the DHT sensor to work?

Yes Uri, it is returning this value:
Humidity: nan
Temperature: nan

with esp32 nodemcu it's ok

**Comment by urish** [comment]
> Yes Uri, it is returning this value:

I'm not sure why - but if you move it to pin 11 it seems to work correctly

**Comment by urish** [comment]
More news: I2C is now working on the Franzininho WiFi!

Sample project: 
* [Franzininho WiFi + LCD1602 over I2C Clock](https://wokwi.com/arduino/projects/323796775459619410)

**Comment by arcostasi** [comment]
That's great! I will share it with the community! Thanks

**Comment by urish** [comment]
Progress update: HTTPS is now working on Arduino Core!

I made a quick example that fetches jokes from an online REST API and displays them on an LCD screen:

https://wokwi.com/arduino/projects/324148356078109268

I've finished the AES/SHA accelerators. Once I finish the RSA accelerator, I hope we'll also get HTTPS on CircuitPython.

**Comment by arcostasi** [comment]
@urish Is it possible to send printf output to serial monitor using ESP-IDF like ESP32 devkit v1?
- [ESP-IDF Example](https://wokwi.com/arduino/projects/324615433106752083)

**Comment by urish** [comment]
@arcostasi yes - but you need to adjust the connection in diagram.json to read:

```
    [ "esp:TX", "$serialMonitor:RX", "", [] ],
    [ "esp:RX", "$serialMonitor:TX", "", [] ],
```

(instead of TX0 / RX0)

**Comment by arcostasi** [comment]
Thanks Uri, printf() is working perfectly! :)

**Comment by urish** [comment]
Excellent!

What do you think is still missing before we can release the Franzininho WiFi board?

**Comment by arcostasi** [comment]
I will do more tests today and give feedback! for now it's ok

**Comment by urish** [comment]
Thank you! 

From my end: need to create a documentation page, add it to the list of "New project" options, create some basic examples (perhaps we can use the examples that you created?)

Nice-to-have: Finish CircuitPython support (so it won't start in Safe mode); MicroPython support

**Comment by arcostasi** [comment]
> From my end: need to create a documentation page, add it to the list of "New project" options, create some basic examples (perhaps we can use the examples that you created?)

Feel free to use my examples! I will create more examples during my tests.

> Nice-to-have: Finish CircuitPython support (so it won't start in Safe mode); MicroPython support

Yes, I believe CircuitPython will be used more by the Franzininho community.

Thanks Uri

**Comment by arcostasi** [comment]
@urish, I'm testing the inputs, in your example of the piano with Franzininho WiFi, when I release a key and press another the input doesn't respond, I need to press it again to hear the buzzer.

- [Mini piano example](https://wokwi.com/arduino/projects/324676273947279956)

**Comment by urish** [comment]
Thanks @arcostasi!

Can you connect the logic analyzer and check if the data is fed in (but not played but the speaker) or if it skips feeding in the data?

On a different note - worked on CircuitPython today. It now runs your code automatically:

https://wokwi.com/arduino/projects/313606939786347074

However, there's one annoying thing: you don't see all the output immediately, as it takes a few microseconds for the USB to connect. This means that if you have a syntax error in your code, you'll need to press "Ctrl+D" to see it.

**Comment by arcostasi** [comment]
Hi @urish,  apparently with the logic analyzer it's normal! I pressed 3 times alternately between the red and orange button and I did not hear the speaker, when I pressed 2 times on the same button without alternating I can hear the speaker.
![logic-analyzer-](https://user-images.githubusercontent.com/5041497/155901934-8217e5d6-14a4-44df-a2e7-0784ebb999ba.png)

**Comment by urish** [comment]
Thanks! Can you please also connect the speaker pin to the logic analyzer?

**Comment by arcostasi** [comment]
> However, there's one annoying thing: you don't see all the output immediately, as it takes a few microseconds for the USB to connect. This means that if you have a syntax error in your code, you'll need to press "Ctrl+D" to see it.

That is great! as soon as I have some free time I'll have fun with CircuitPython.

**Comment by arcostasi** [comment]
> Thanks! Can you please also connect the speaker pin to the logic analyzer?

A square wave appeared in the speaker channel (D2), but the audio did not play in my headphones.

![logic-analyzer-2](https://user-images.githubusercontent.com/5041497/155902237-c6d00dbb-117a-4053-a3f1-10f3b50e519c.png)

**Comment by urish** [comment]
Thanks! So I suspect an issue with the buzzer. Does the frequency of the square wave look correct? (you can ask PulseView to calculate the frequency by holding Shift and selecting the region between two adjacent peaks (low-to-high transitions).

Also, do you observe any difference when the buzzer is set to [accurate mode](https://docs.wokwi.com/parts/wokwi-buzzer#operation-modes)?

**Comment by arcostasi** [comment]
In accurate mode I can hear the sound with noises, even switching between the buttons.

![analyser](https://user-images.githubusercontent.com/5041497/155902769-a1d16273-10d5-4b2d-be6a-9308971dc59e.png)

**Comment by urish** [comment]
Thanks @arcostasi. I think I found the issue with the smooth mode - it looked at edge-to-edge time instead of peak-to-peak time to determine the frequency. The ESP32 uses a non-50% duty cycle, so it confused the smooth mode algorithm:

![image](https://user-images.githubusercontent.com/892318/155964474-05846b85-8c08-4501-8c63-9e898fc964b8.png)

You can see that the frequency is right, but the duty cycle (% of the time the signal is high in each cycle) is not 50%. This is different from how Arduino does PWM. 

I have updated the algorithm, can you give it another go please?

**Comment by arcostasi** [comment]
Hi @urish, it worked perfectly now! Thank you

**Comment by arcostasi** [comment]
@urish, I did another test with Franzininho WiFi, was the analog reading implemented? Thanks

Here is an example: [RGB LED with Slide Potentiometer](https://wokwi.com/arduino/projects/324682033130373716)

**Comment by urish** [comment]
> was the analog reading implemented?

As far as I can tell not yet

**Comment by arcostasi** [comment]
Ok, no problem! Thank you

**Comment by urish** [comment]
Update: analogRead (ADC) implemented!

**Comment by arcostasi** [comment]
Thanks @urish, worked perfectly!

**Comment by arcostasi** [comment]
Here's an example of using a [servo with Franzininho WiFi](https://wokwi.com/projects/327061759288410708)

**Comment by urish** [comment]
Documentation for the Franzininho WiFi board: https://docs.wokwi.com/parts/board-franzininho-wifi

**Comment by urish** [comment]
The last missing feature is now ready: SPI support. Here's a quick demo:

https://wokwi.com/projects/329013233501340242

So now we have feature parity with the ESP32 and can declare the Franzininho WiFi simulation ready. Thanks for everyone who supported this feature :-)

**Comment by urish** [comment]
Franziniho WiFi now has a "New project" page: https://wokwi.com/projects/new/franzininho-wifi
